### ISSUE TITLE
Feature/implementation support request: cluster graph integration

### ISSUE THREAD

**Issue by WillemMali** [kubernetes/dashboard#2008]
I deploy K8s/`dashboard` via `kops`, and I would like to integrate  [`kubernetes-ui/graph`](https://github.com/kubernetes-ui/graph) into the dashboard. It would show up on the condition that `cluster-insight` is running.

Is there currently a plugin/component system for `dashboard`? Could someone who has architectural overview of `dashboard` help me make a rough outline of the work that has to be done to integrate the cluster graph into `dashboard`?

**Comment by bryk** [comment]
There's no extension mechanism at the moment. So this is all do-it-yourself. I think, though, this is the right moment to start thinking about the extensions framework for UI for real.

**Comment by bryk** [comment]
@floreks @maciaszczykm @cheld WDYT of doing this via one-off compile time hardcoding versus extensions framework?

**Comment by cheld** [comment]
The scope plugin uses an Iframe. https://github.com/kubernetes/dashboard/pull/1949

The helm plugin we are thinking to integrate directly. Not sure about the best approach

In addition @floreks is thinking about a global settings page, baked by a config map to enable/disable plugins - like in chrome

**Comment by bryk** [comment]
Yeah. So there's a lot of these things appearing recently. I expect to see more. We could tackle this problem in a generic way to help everyone in one sweep. This is a large project, though.

**Comment by floreks** [comment]
It would be nice to have some extensions framework, but there are many things to consider. We need to discuss that first and think how to provide such a mechanism in a generic and dynamic way that will not require direct dashboard changes.

What I'm working on is not a plugin mechanism but integrations framework for our backend. This means that still we will need to introduce all integrations related code upstream. It will be just easier to integrate with dashboard.

**Comment by bryk** [comment]
Do you have any designs for the integrations framework already? Can we talk about this? :)

**Comment by floreks** [comment]
It's still an early sketch. I've been preparing other parts of backend for such framework. I want to minimize work needed to introduce new integrations later on so API still changes. Also I need to introduce some kind of storage provider with storage sync that will keep config up to date and integrate that with integrations manager.

There are many things to consider so I'm trying not to rush anything. It would be nice to discuss it. We can do that tomorrow.

**Comment by WillemMali** [comment]
Hi, just checking in on the extension framework/cluster graph integration, is there anything I could do to help further this integration?

@floreks, could you perhaps write a roadmap of what you think the issues to discuss are before implementing a extension framework?

**Comment by floreks** [comment]
First part can be seen in https://github.com/kubernetes/dashboard/pull/2017 pull request. I still need to clean up some code but overall structure is already there. Right now I don't have clear roadmap. For sure after this part I will be working on some storage sync to be able to save per user config.

Unfortunately right now I don't see any tasks that could be done in parallel. Base part has to finished first and then new integrations i.e. prometheus for metric can be added.
